indico/modules/events/registration/wallets/apple.py

```diff
@@ -0,0 +1,144 @@
+        ticket.addBackField('back-registration-name', registration.full_name, _('Name'))
+        ticket.addBackField('back-ticket-number', f'#{registration.friendly_id}', _('Ticket number'))
+        ticket.addBackField('back-event-date', event_date, _('Date'))
+        ticket.addBackField('back-registration-email', registration.email, _('Email'))
+        ticket.addBackField('back-event-url', f'<a href="{self.event.external_url}">{_("Event page")}</a>', _('Link'))
+        reg_url = url_for('event_registration.display_regform', registration.locator.uuid, _external=True)
+        ticket.addBackField('back-registration-details', f'<a href="{reg_url}">{_("Registration details")}</a>',
+                            _('Link'))
+
+        signals.event.registration.apple_wallet_ticket_object.send(self.event, obj=ticket)
+        return ticket
+
+    def build_pass_object(self, registration):
+        from indico.modules.events.registration.util import get_persons, get_ticket_qr_code_data
+
+        # Extract Certificate details from certificate.pem itself
+        self.cert = x509.load_pem_x509_certificate(self.settings['apple_wallet_certificate'].encode())
+        cert_details = {d.split('=')[0]: d.split('=')[1] for d in self.cert.subject.rfc4514_string().split(',')}
+        passfile = IndicoPass(self.build_ticket_object(registration), passTypeIdentifier=cert_details['UID'],
+                              organizationName=cert_details['O'], teamIdentifier=cert_details['OU'])
+        passfile.logoText = cert_details['O']  # Add Organization name at the top
+        passfile.backgroundColor = '#007cac'
+        passfile.foregroundColor = passfile.labelColor = '#ffffff'
+        passfile.serialNumber = registration.ticket_uuid
+        qr_data = get_ticket_qr_code_data(get_persons([registration])[0])
+        passfile.barcode = Barcode(message=json.dumps(qr_data, separators=(',', ':')), format=BarcodeFormat.QR)
+
+        # Including the icon and logo is necessary for the passbook to be valid.
+        if config.WALLET_LOGO_URL:
+            passfile.add_file_from_url('icon.png', config.ABSOLUTE_WALLET_LOGO_URL)
+            passfile.add_file_from_url('logo.png', config.ABSOLUTE_WALLET_LOGO_URL)
+        else:
+            passfile.add_default_images()
+        signals.event.registration.apple_wallet_object.send(self.event, obj=passfile)
+        return passfile
+
+    def get_ticket(self, registration) -> str:
+        passfile = self.build_pass_object(registration)
+        wwdr_path = os.path.join(current_app.root_path, 'modules', 'events', 'registration', 'wallets',
+                                 'apple-wwdr.pem')
+        # Create and output the Passbook file (.pkpass)
+        temp = tempfile.NamedTemporaryFile(prefix='apple_wallet_', dir=config.TEMP_DIR, delete=False)
+        registration.apple_wallet_serial = passfile.serialNumber  # Save ticket serial for updates
```

Comment:
I'd like to have this closer to where we use it above. How about a `_get_wallet_serial` method which:

- returns the `apple_wallet_serial` if there is one
- otherwise assigns it (using the ticket uuid) and then returns it

because the current code here looks like it needs to set the serial every time, even though it does not change

Reply:
I changed it slightly to avoid reassigning it every time (even though that's a no-op and not actually triggering a DB operation), but it was very close to what I had in mind.